src/appleseed/renderer/kernel/lighting/tracer.h

```diff
@@ -79,34 +79,52 @@ class Tracer
     // with the closest fully opaque occluder and the transmission factor up
     // to (but excluding) this occluder, or a miss if there is no fully opaque
     // occluder in this direction.
-    template <typename ReturnType = void>
-    ReturnType trace(
+    void trace(
         const ShadingContext&       shading_context,
-        const ShadingPoint&         origin,
         const ShadingRay&           ray,
         Spectrum&                   transmission);
 
-    template <typename ReturnType = void>
-    ReturnType trace(
+    void trace(
         const ShadingContext&       shading_context,
         const ShadingRay&           ray,
+        Spectrum&                   transmission,
+        const ShadingPoint*&        shading_point);
```

Comment:
The reference itself is not const, it's the pointer that is. This is one way to return a pointer from a function. Another way would be to pass a pointer to a pointer to a ShadingPoint: `const ShadingPoint**`